articles/functional-programming-in-java/index.md

```diff
@@ -0,0 +1,225 @@
+
+```
+
+Since the method body only contains a return statement, we can simplify it even further:
+
+```java
+
+num -> num % 2 == 0
+
+```
+
+At this point, you may be wondering why we omit the types of the parameter and returned value. With Java being a statically typed language, it may not be very intuitive to you why this is the case. Yet, lambda expressions are strongly and statically typed, even though we don't show types. We will explain this in detail later.
+
+### Lambda Expressions vs Anonymous Inner Classes
+Lambda expressions, being a way to create "methods" in-line, draws some comparison to anonymous inner classes. As you may recall, anonymous inner-classes provide a way to build objects on the fly for use as arguments. We often use anonymous inner classes to pass event listeners into swing components:
+
+```java
+
+JButton b1 = new JButton("Hello World");
+
+b1.addActionListener(new ActionListener() {
+
+    public void actionPerformed(ActionEvent e) {
+
+        System.out.println("This is pretty cool");
+
+    }
+
+});
+
+```
+
+With the anonymous inner-class, we create the class in-line as it lacks any re-usability. Lambda expressions work almost the same way; they are also used to create single-use objects in-line. These objects are essentially instances of wrapper classes for the methods we use. Behind the scenes, lambda expressions are implementations of what is called `functional interfaces`.
+
+#### Functional Interfaces
+Functional interfaces are essentially interfaces that contain one unimplemented method. This allows the compiler to map lambda expressions into them. The idea is that they represent a specific functionality and not some kind of entity.
+
+If you carefully read that description, you will notice that the ActionListener interface follows this. It has only one method to implement and represents a functionality, i.e., a response to an action. As such, we may instead express it as a lambda expression:
+
+```java
+
+JButton b1 = new JButton("Hello World");
+
+b1.addActionListener(event -> {
+
+    System.out.println("This is pretty cool");
+
+});
+
+```
+
+Since there is only one method to implement, the only meaningful code is its implementation. With the lambda expression being only the implementation, we can use it to reduce boilerplate code.
+
+This leads back to the question about type safety. In our example, the addActionListener method requires an ActionListener implementation. With this, the compiler knows the lambda expression's parameter and return types. After all, the ActionListener interface has one method, which must be the method we are implementing. Since it knows what method we are trying to implement, it can know the types from the interface's declaration. This allows us to omit the parameter and return types, unlike traditional methods.
+**Note: _functional interfaces often have their argument and return types as generics. In this case, the method accepting it specifies the types._**
```

Comment:
This is one of the formatting styles I was telling you to look at. Also, use backticks to highlight the keywords or words of interest.

Reply:
May you please explain exactly what the issue is with the formatting styles in this section? I do not seem to understand what it is.

Reply:
It was to show how you can style some parts of your articles. Parts like notes can be in bold to separate them from the normal statements. You can also highlight keywords to ensure they also stand out.